docs/intermediate/configure-php-xdebug.rst
.. include:: /_includes/snippets/__ANNOUNCEMENTS__.rst

.. _configure_php_xdebug:

********************
Configure PHP Xdebug
********************

This section explains in depth how to enable and use PHP Xdebug with the Blibox.


**Table of Contents**

.. contents:: :local:


Introduction
============

In order to have a working Xdebug, you need to ensure two things:

1. **PHP Xdebug** must be configured and enabled in PHP itself
2. Your **IDE/editor** must be configured and requires a way talk to PHP

Configuring PHP Xdebug will slightly differ when configuring it for a dockerized environment.
This is due to the fact that Docker versions on different host os have varying implementations of
how they connect back to the host.

Most IDE or editors will also require different configurations for how they talk to PHP Xdebug.
This is at least most likely the case for ``xdebug.idekey``.

.. seealso:: :ref:`configure_php_xdebug_options`


Configure Xdebug
================

Docker on Linux
---------------

Docker on Linux allows Xdebug to automatically connect back to the host system without the need
of an explicit IP address.

.. toctree::
   :glob:
   :maxdepth: 1
   :hidden:

   /intermediate/configure-php-xdebug/linux/*

.. seealso::

   * :ref:`configure_php_xdebug_lin_atom`
   * :ref:`configure_php_xdebug_lin_phpstorm`
   * :ref:`configure_php_xdebug_lin_sublime`
   * :ref:`configure_php_xdebug_lin_vscode`

Docker on MacOS
---------------

.. toctree::
   :glob:
   :maxdepth: 1
   :hidden:

   /intermediate/configure-php-xdebug/macos/*

.. seealso::

   * :ref:`configure_php_xdebug_mac_atom`
   * :ref:`configure_php_xdebug_mac_phpstorm`
   * :ref:`configure_php_xdebug_mac_sublime`
   * :ref:`configure_php_xdebug_mac_vscode`

Docker on Windows
-----------------

.. toctree::
   :glob:
   :maxdepth: 1
   :hidden:

   /intermediate/configure-php-xdebug/windows/*

.. seealso::

   * :ref:`configure_php_xdebug_win_atom`
   * :ref:`configure_php_xdebug_win_phpstorm`
   * :ref:`configure_php_xdebug_win_sublime`
   * :ref:`configure_php_xdebug_win_vscode`

Docker Toolbox
--------------

Docker Toolbox configuration is equal, no matter if it is started on MacOS or Windows, as both
use a Linux system inside VirtualBox.

.. toctree::
   :glob:
   :maxdepth: 1
   :hidden:

   /intermediate/configure-php-xdebug/toolbox/*

.. seealso::

   * :ref:`configure_php_xdebug_toolbox_atom`
   * :ref:`configure_php_xdebug_toolbox_phpstorm`
   * :ref:`configure_php_xdebug_toolbox_sublime`
   * :ref:`configure_php_xdebug_toolbox_vscode`


docs/intermediate/add-custom-environment-variables.rst
.. include:: /_includes/snippets/__ANNOUNCEMENTS__.rst

.. _add_custom_environment_variables:

********************************
Add custom environment variables
********************************

If your application requires a variable to determine if it is run under development or
production, you can easily add it and make PHP aware of it.


**Table of Contents**

.. contents:: :local:


Add custom environment variables
================================

This is fairly simple. Any variable inside the ``.env`` file is considered an environment variable
and automatically known to PHP.

If you for example require a variable ``APPLICATION_ENV``, with a value of ``production``, you
would add the following to the ``.env`` file:

.. code-block:: bash
   :caption: .env
   :emphasize-lines: 1

   APPLICATION_ENV=production

You need to restart the Blibox for the changes to take effect.

.. note::
   There is already a proposed section inside the ``.env`` file at the very bottom
   to add you custom variables to differentiate them from the Blibox required variables.


Use custom environment variables
================================

Accessing the above defined environment variable on the PHP side is also fairly simple.
You can use the PHP's built-in function ``getenv`` to obtain the value:

.. code-block:: php
   :caption: index.php
   :emphasize-lines: 3

   <?php
   // Example use of getenv()
   echo getenv('APPLICATION_ENV');
   ?>


CHANGELOG.md
# Changelog

Make sure to have a look at [UPDATING](https://github.com/cytopia/blibox/blob/master/UPDATING.md) to see any required steps for updating
major versions.


## Unreleased


## Release v1.7.0 (2020-03-24)

#### Added
- Python Flask


## Bugfix Release v1.6.3 (2020-03-23)

#### Fixed
- Fixed PHP FPM images
- Fixed cert-gen for HAProxy
- Various spelling errors in documentation
- PostgreSQL startup without a password
#### Added
- [#686](https://github.com/cytopia/blibox/pull/686) Added documentation for ExpressEngine
- New .env var: `PGSQL_HOST_AUTH_METHOD`


## Bugfix Release v1.6.2 (2020-02-06)

#### Added
- [#670](https://github.com/cytopia/blibox/issues/670) Add `phalcon` binary to PHP 7.3 and 7.4
- [#664](https://github.com/cytopia/blibox/issues/664) Add PHP `yaml` extension


## Bugfix Release v1.6.1 (2020-01-05)

#### Fixed
- [#662](https://github.com/cytopia/blibox/issues/662) Update to latest Symfony CLI


## Release v1.6.0 (2020-01-04)

#### Changed
- [#642](https://github.com/cytopia/blibox/issues/642) Make email catch-all configurable
- [#265](https://github.com/cytopia/blibox/issues/265) Make SSL vhost settings configurable

#### Added
- [#615](https://github.com/cytopia/blibox/issues/615) Add phpmd
- [#378](https://github.com/cytopia/blibox/issues/378) Allow to mount local `.ssh/` directory into PHP container (read-only)


## Release v1.5.0 (2020-01-03)

#### Added
- [#654](https://github.com/cytopia/blibox/issues/654) Added Opcache Control Panel
- Integration tests for MySQL Docker image


## Release v1.4.0 (2020-01-02)

#### Fixed
- [#618](https://github.com/cytopia/blibox/issues/618) Update Compose version to 2.3
- [#614](https://github.com/cytopia/blibox/issues/614) Update to latest mhsendmail binary with `-o` flag
- [#265](https://github.com/cytopia/blibox/issues/265) Fix http to https redirect for projects

#### Changed
- [#642](https://github.com/cytopia/blibox/issues/642) Be able to disable email catch-all and still run postfix


## Release v1.3.0 (2019-12-29)

#### Fixed
- [#626](https://github.com/cytopia/blibox/issues/626) Cannot create MongoDB database with Adminer

#### Added
- latest PHP-FPM images
- PHP module: solr
- PHP module: ssh2
- phpMyAdmin to 5.0.0

#### Changed
- Use official PHP 7.4 Docker image as base
- Updated Adminer to 4.7.5
```
sed -i'' 's/^<?php$/<?php if(!function_exists("get_magic_quotes_runtime")){function get_magic_quotes_runtime(){return false;}}if(!function_exists("get_magic_quotes_gpc")){function get_magic_quotes_gpc(){return false;}}/g' adminer-4.7.5-en.php
sed -i'' 's/while(list(\$y,\$X)=each(\$qg))/foreach ($qg as $y => $x)/g' adminer-4.7.5-en.php
sed -i'' 's/error_reporting(6135)/error_reporting(0)/g' adminer-4.7.5-en.php
```
- Updated phpMyAdmin to 4.9.3


## Release v1.2.0 (2019-12-01)

#### Fixed
- [#622](https://github.com/cytopia/blibox/issues/622) Certificate Generation Settings
- [#640](https://github.com/cytopia/blibox/issues/640) Mac OS Catalina invalidates virtual host certificates
- [#592](https://github.com/cytopia/blibox/issues/592) sqlsrv connection problem

#### Added
- Added latest PHP-FPM images
- Added latest MySQL images
- Added latest Apache/Nginx images

#### Changed
- Made PHP 7.3 the default version


## Release v1.1.0 (2019-11-24)

#### Fixed
- [#644](https://github.com/cytopia/blibox/issues/644) Fix GD jpeg support missing in PHP 7.4
- [#619](https://github.com/cytopia/blibox/issues/619) Fix PHP 7.x  WebP support with imagick

#### Added
- Make Ngrok region configurable via `NGROK_REGION` env variable
- [#641](https://github.com/cytopia/blibox/issues/641) Added phpPgAdmin v7 for PHP >=7
- [#594](https://github.com/cytopia/blibox/issues/594) Tool: ghostscript
- Extensive GitHub Action CI checks
- Added new PostgreSQL images
- Added new MongoDB images

#### Changed
- Updated PHP Docker images (PHP version, modules and tools)


## Bugfix Release v1.0.2 (2019-05-21)

#### Fixed
- Fixed various typos in documentation
- Fix CI tests: They still expected a mounted mail directory instead of a Docker volume
- Fix mods for PHP 8.0
- Make npm binaries available in $PATH
- Remove orphaned mentions of HOST_PATH_MYSQL_DATADIR
- Allow symlinks in autostart scripts

#### Added
- Documentation: Setup ProcessWire
- Tool: Angular CLI
- Tool: Laravel Lumen
- Tool: [prestissimo](https://github.com/hirak/prestissimo)
- Tool: [yq](https://github.com/mikefarah/yq)
- Module: OAuth

#### Removed
- Enchant module for PHP 7.4 and PHP 8.0 (build breaks)


## Bugfix Release v1.0.1 (2019-03-24)

This is a bugfix release and everybody is encouraged to upgrade to this tag as soon as possible.
No explicit actions to be taken for updating.

#### Fixed
- [#373](https://github.com/cytopia/blibox/issues/373) Read custom MySQL configuration
- [#540](https://github.com/cytopia/blibox/issues/540) Fix '&' password substitution in mysqldump-secure
- [#209](https://github.com/cytopia/blibox/issues/209) Documentation for Xdebug on Docker for Windows

#### Changed
- Updated MongoDB cli tools in PHP image
- Updated PostgreSQL cli tools in PHP image
- MySQL images are now bound to a specific Docker tag and are built nightly
- [#506](https://github.com/cytopia/blibox/issues/506) Documentation improvements for connecting to databases

#### Added
- [#536](https://github.com/cytopia/blibox/issues/536) Added @vue/cli and @vue/cli-service-global



## Release v1.0.0 (2019-03-19)

This is the first major stable release of the Blibox.

#### Changed
- Everything from v1.0.0-alpha1 has been backported
- Everything from v0.15.0 has been backported



## Pre-Release v1.0.0-alpha1 (2019-03-09)

#### Changed
- Use Docker volumes instead of directory mounts for stateful data (MySQL, PgSQL and MongoDB)
    - This fixes various mount issues on Windows: #175 #382
    - This improves general performance
- Use Official MySQL, MariaDB and Percona Docker container



## Release v0.15.0 (2019-03-09)

This will be the last v0.x release.

#### Fixed

.blibox/www/config.php
<?PHP
// Measure time
$TIME_START = microtime(true);

// Start session
session_start();

// Turn on all PHP errors
error_reporting(-1);


// Shorten DNS timeouts for gethostbyname in case DNS server is down
putenv('RES_OPTIONS=retrans:1 retry:1 timeout:1 attempts:1');


$BLIBOX_VERSION = 'v1.7.0';
$BLIBOX_DATE = '2020-03-24';
$BLIBOX_API_PAGE = 'blibox-api/status.json';

//
// Set Directories
//
$CONF_DIR	= dirname(__FILE__);
$LIB_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'lib';
$VEN_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'vendor';
$LOG_DIR	= dirname(dirname($CONF_DIR)) . DIRECTORY_SEPARATOR . 'log' . DIRECTORY_SEPARATOR . 'blibox';


//
// Load Base classes
//
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseClass.php';
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseInterface.php';



//
// Set Docker addresses
//
$DNS_HOST_NAME		= 'bind';
$PHP_HOST_NAME		= 'php';
$HTTPD_HOST_NAME	= 'httpd';
$MYSQL_HOST_NAME	= 'mysql';
$PGSQL_HOST_NAME	= 'pgsql';
$REDIS_HOST_NAME	= 'redis';
$MEMCD_HOST_NAME	= 'memcd';
$MONGO_HOST_NAME	= 'mongo';


//
// Lazy Container Loader
//
function loadFile($class, $base_path) {
	static $_LOADED_FILE;

	if (isset($_LOADED_FILE[$class])) {
		return;
	}

	require $base_path . DIRECTORY_SEPARATOR . $class . '.php';
	$_LOADED_FILE[$class] = true;
	return;
}
function loadClass($class) {

	static $_LOADED_LIBS;

	if (isset($_LOADED_LIBS[$class])) {
		return $_LOADED_LIBS[$class];
	} else {

		$lib_dir = $GLOBALS['LIB_DIR'];
		$cnt_dir = $GLOBALS['LIB_DIR'] . DIRECTORY_SEPARATOR . 'container';

		switch($class) {
			//
			// Lib Classes
			//
			case 'Logger':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Logger::getInstance();
				break;

			case 'Html':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Html::getInstance();
				break;

			case 'Helper':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Helper::getInstance();
				break;

			//
			// Docker Container Classes
			//
			case 'Php':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Php::getInstance($GLOBALS['PHP_HOST_NAME']);
				break;

			case 'Dns':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Dns::getInstance($GLOBALS['DNS_HOST_NAME']);
				break;

			case 'Httpd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Httpd::getInstance($GLOBALS['HTTPD_HOST_NAME']);
				break;

			case 'Mysql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Mysql::getInstance($GLOBALS['MYSQL_HOST_NAME'], array(
					'user' => 'root',
					'pass' => loadClass('Helper')->getEnv('MYSQL_ROOT_PASSWORD')
				));
				break;

			case 'Pgsql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Pgsql::getInstance($GLOBALS['PGSQL_HOST_NAME'], array(
					'user' => loadClass('Helper')->getEnv('PGSQL_ROOT_USER'),
					'pass' => loadClass('Helper')->getEnv('PGSQL_ROOT_PASSWORD'),
					'db' => 'postgres'
				));
				break;

			case 'Redis':

				// Check if redis is using a password
				$REDIS_ROOT_PASSWORD = '';

				$_REDIS_ARGS = loadClass('Helper')->getEnv('REDIS_ARGS');
				$_REDIS_PASS = preg_split("/--requirepass\s+/",  $_REDIS_ARGS);
				if (is_array($_REDIS_PASS) && count($_REDIS_PASS)) {
					// In case the option is specified multiple times, use the last effective one.
					$_REDIS_PASS = $_REDIS_PASS[count($_REDIS_PASS)-1];
					if (strlen($_REDIS_PASS) > 0) {
						$REDIS_ROOT_PASSWORD = $_REDIS_PASS;
					}
				}

				loadFile($class, $cnt_dir);
				if ($REDIS_ROOT_PASSWORD == '') {
					$_LOADED_LIBS[$class] = \blibox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME']);
				} else {
					$_LOADED_LIBS[$class] = \blibox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME'], array(
						'pass' => $REDIS_ROOT_PASSWORD,
					));
				}
				break;

			case 'Memcd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Memcd::getInstance($GLOBALS['MEMCD_HOST_NAME']);
				break;

			case 'Mongo':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Mongo::getInstance($GLOBALS['MONGO_HOST_NAME']);
				break;

			// Get optional docker classes
			default:
				// Unknown class
				exit('Class does not exist: '.$class);
		}
		return $_LOADED_LIBS[$class];
	}
}
